packages/next/src/lib/metadata/resolve-metadata.tsx

```diff
@@ -0,0 +1,243 @@
+          parent: ResolvingMetadata
+        ) => Promise<Metadata>
+      }>
+      path?: string
+    }
+
+// Merge the source metadata into the resolved target metadata.
+function merge(
+  target: ResolvedMetadata,
+  source: Metadata,
+  templateStrings: {
+    title: AbsoluteTemplateString
+    openGraphTitle: AbsoluteTemplateString
+    twitterTitle: AbsoluteTemplateString
+  }
+) {
+  let updatedStashedTitle: AbsoluteTemplateString = templateStrings.title
+  let updatedStashedOpenGraphTitle: AbsoluteTemplateString =
+    templateStrings.openGraphTitle
+  let updatedStashedTwitterTitle: AbsoluteTemplateString =
+    templateStrings.twitterTitle
+
+  for (const key_ in source) {
+    const key = key_ as keyof Metadata
+
+    switch (key) {
+      case 'other': {
+        target.other = { ...target.other, ...source.other }
+        break
+      }
+      case 'title': {
+        updatedStashedTitle = resolveTitle(templateStrings.title, source.title)
+        target.title = updatedStashedTitle
+        break
+      }
+      case 'openGraph': {
+        if (typeof source.openGraph !== 'undefined') {
+          target.openGraph = {
+            ...resolveOpenGraph(source.openGraph),
+          }
+          if (source.openGraph && 'title' in source.openGraph) {
+            updatedStashedOpenGraphTitle = resolveTitle(
+              templateStrings.openGraphTitle,
+              source.openGraph.title
+            )
+            target.openGraph.title = updatedStashedOpenGraphTitle
+          }
+        } else {
+          target.openGraph = null
+        }
+        break
+      }
+      case 'twitter': {
+        if (typeof source.twitter !== 'undefined') {
+          target.twitter = { ...source.twitter }
+          if (source.twitter && 'title' in source.twitter) {
+            updatedStashedTwitterTitle = resolveTitle(
+              templateStrings.twitterTitle,
+              source.twitter.title
+            )
+            target.twitter.title = updatedStashedTwitterTitle
+          }
+        } else {
+          target.twitter = null
+        }
+        break
+      }
+      default: {
+        // TODO: Make sure the type is correct.
+        // @ts-ignore
+        target[key] = source[key]
+        break
+      }
+    }
+  }
+
+  return {
+    title: updatedStashedTitle,
+    openGraphTitle: updatedStashedOpenGraphTitle,
+    twitterTitle: updatedStashedTwitterTitle,
+  }
+}
+
+export async function resolveMetadata(metadataItems: Item[]) {
+  const resolvedMetadata = createDefaultMetadata()
+
+  const committedTitle: AbsoluteTemplateString = {
+    absolute: '',
+    template: '',
+  }
+  const committedOpenGraphTitle: AbsoluteTemplateString = {
+    absolute: '',
+    template: '',
+  }
+  const committedTwitterTitle: AbsoluteTemplateString = {
+    absolute: '',
+    template: '',
+  }
+  let stashedTitles
+
+  // from root layout to page metadata
+  for (let i = 0; i < metadataItems.length; i++) {
+    const item = metadataItems[i]
```

Comment:
```suggestion

  let lastLayer = 0;
  // from root layout to page metadata
  for (let i = 0; i < metadataItems.length; i++) {
    const item = metadataItems[i]
    if (item.layer > lastLayer) {
      stashedTitleTemplate = resolvedMetadata.title.template || '';
      // ... other templates...
    }
    lastLayer = item.layer
```

stash here when we enter a new layer. Also, we only need to stash the template not the whole title. this prevents us from needing to clone the title to avoid it mutating later